### PR TITLE
Support for gathering and exporting Prometheus metrics

### PR DESCRIPTION
**Pull request by rossabaker**
This is a new feature, but there's nothing binary-incompatible about it, so proposing it for 0.15.x.

### CODE REVIEW

**Comment by rossabaker** [comment]
Fixes #784

File: prometheus-server/src/main/scala/org/http4s/prometheus/server/Metrics.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def methodLabel(method: Method) =
+      if (whitelistedMethods.contains(method)) method.name
+      else "other"
+
+    Service.lift { req: Request =>
+      activeRequests.inc()
+      val start = System.nanoTime
+      val future = service(req).get.map { result =>
+        val stop = System.nanoTime
+        activeRequests.inc(-1.0)
```
Comment (bryce-anderson) [inline prometheus-server/src/main/scala/org/http4s/prometheus/server/Metrics.scala]: Is this going to be misleading? As I understand it, this is only the head of the response and the body may take a while to render after this.

Comment (rossabaker) [inline prometheus-server/src/main/scala/org/http4s/prometheus/server/Metrics.scala]: Yes.  This should be in record.

Comment (bryce-anderson) [inline prometheus-server/src/main/scala/org/http4s/prometheus/server/Metrics.scala]: I believe even record doesn't care about the body.

**Comment by rossabaker** [comment]
There was concern about duplication between the Dropwizard and Prometheus implementations.  I've also never been hugely happy with the middleware implementation of this, as it misses times and status changes that happen in the server backend.  I have a branch sitting around that hooks into a blaze stage that could address these concerns, with the idea that logging and metrics are already well handled in the servlet containers.

If people really want this, we can proceed.  But I think we can do better.

**Comment by MoAllenBellevue** [comment]
We would like to use Prometheus metrics in an http4s app that we are building.  Are there any plans to revive this branch and merge it?  If not, any suggestions for alternative approaches? Thanks.

**Comment by rossabaker** [comment]
I doubt we would want to support this as is over the long term.  I think a better approach to metrics is not a middleware, but to wrap around the outer layers of the server.  It's a portability vs. accuracy tradeoff.

I'm also not using Prometheus anywhere at the moment, so I've lost personal interest on that.  But I know several of our users do, so we should have some solution.

There is a [Dropwizard to Prometheus bridge](https://www.robustperception.io/exposing-dropwizard-metrics-to-prometheus/) that I've never tried, but should play nicely with the existing metrics module.

**Comment by rossabaker** [comment]
Superseded by #1684

**Pull request closed without merging**